File: svn/branches/game/CaseyPyroJ/src/edu/wpi/first/wpilibj/templates/HAL.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package edu.wpi.first.wpilibj.templates;

import edu.wpi.first.wpilibj.*;
//import edu.wpi.first.addons.*;

/**
 *
 * @author Currey-WilsonC
 */
public class HAL implements Part
{
    private int tmr;

    public Joystick rStick;
    public Joystick lStick;
    public Joystick hStick;

    private AnalogChannel kickerAnalogChannel;
    //Left motors currently thought to be positive=forward.
    private CANJaguar frMotor;
    private CANJaguar flMotor;
    private CANJaguar brMotor;
    private CANJaguar blMotor;
    private CANJaguar winchShootMotor;

    //Relay winchBrake;
    /*CANJaguar winchHangMotor; CANJaguar winchHangMotor2; */ // hanger is dead
    private CANJaguar suckerMotor;
    private CANJaguar suckerMotor2;
    private Compressor compressor;

    private DigitalInput kickLimit;
    private Solenoid kickerSolenoid, kickerForwardSolenoid;
    //private Encoder frEncoder, flEncoder,brEncoder, blEncoder;

    private DriverStationLCD dsLCD;
    //private Servo armServo;

    private int rStickPort = 3;
    private int lStickPort = 2;
    private int hStickPort = 1;
    private int frMotorPort = 3; private int flMotorPort = 4;
    private int brMotorPort = 9; private int blMotorPort = 2;
    //private int armRelayChannel = 1;
    private int kickerRelayChannel = 3;
    private int winchShootMotorPort = 6;

    private int suckerMotorPort2 = 7;
    private int suckerMotorPort = 8;
    //private int gyroPort = 1;
    //private int winchHangMotorPort = 7; // Retract - negative
    //private int winchHangMotor2Port = 11; // Retract - positive
    //encoder ports subject to change
    /*private int frEncoderPort1 = 3; private int frEncoderPort2 = 4;
    private int flEncoderPort1 = 7; private int flEncoderPort2 = 8;
    private int brEncoderPort1 = 5; private int brEncoderPort2 = 6;
    private int blEncoderPort1 = 1; private int blEncoderPort2 = 2;*/

    private int pressureSwitchChannel = 9;
    private int compressorRelayChannel = 1;

    private int kickerForwardSolenoidChannel = 1;
    private int kickerSolenoidChannel = 2;
    private int kickLimitPort = 10;

    private int kickerAnalogChannelPort = 3;

    private double brMotorPower;
    private double blMotorPower;
    private double frMotorPower;
    private double flMotorPower;
    private double winchShootPower;
    private double winchHangPower;
    private double suckerPower;
    private boolean kickerValue;
    private boolean kickerForwardValue;
    private boolean armValue;
    private boolean kickLimitValue;
    private boolean hangerBreakValue;
    //Relay::Value winchBrakeValue;
    private int scriptMode;

    private boolean shootWinchBreaker;
    private double shootWinchBreakerCurrent = 13;

    public int latchLockedAngle = 100, latchUnlockedAngle = 180;

    public HAL()
    {
        tmr = 0;

        rStick = new Joystick(rStickPort);
        lStick = new Joystick(lStickPort);
        hStick = new Joystick(hStickPort);

        kickerAnalogChannel = new AnalogChannel(kickerAnalogChannelPort);
        frMotor = new CANJaguar(frMotorPort);
        flMotor = new CANJaguar(flMotorPort);
        brMotor = new CANJaguar(brMotorPort);
        blMotor = new CANJaguar(blMotorPort);
        winchShootMotor = new CANJaguar(winchShootMotorPort);
        //winchBrake(winchBrakeRelayChannel),
        //winchHangMotor(winchHangMotorPort),
        //winchHangMotor2(winchHangMotor2Port),
        suckerMotor = new CANJaguar(suckerMotorPort /*,CANJaguar::kCurrent*/);
        suckerMotor2 = new CANJaguar(suckerMotorPort2 /*,CANJaguar::kCurrent*/);
        compressor = new Compressor(pressureSwitchChannel, compressorRelayChannel);
        kickLimit = new DigitalInput(kickLimitPort);
        kickerSolenoid = new Solenoid(kickerSolenoidChannel);
        kickerForwardSolenoid = new Solenoid(kickerForwardSolenoidChannel);
        /*frEncoder = new Encoder(frEncoderPort1, frEncoderPort2, true);
        flEncoder = new Encoder(flEncoderPort1, flEncoderPort2, true);
        brEncoder = new Encoder(brEncoderPort1, brEncoderPort2);
        blEncoder = new Encoder(blEncoderPort1, blEncoderPort2, true);*/
        dsLCD = DriverStationLCD.getInstance();
        //armServo(armServoPort);
        compressor.start();
        brMotorPower = 0;
        blMotorPower = 0;
        frMotorPower = 0;
        suckerPower = 0;
        winchShootPower = 0.0;
        kickerValue = false;
        armValue = hangerBreakValue = false;
        //winchBrakeValue = Relay::kOff;
        scriptMode = 0;
        /*frEncoder.start();
        flEncoder.start();
        brEncoder.start();
        blEncoder.start();*/
        shootWinchBreaker = false;
    }

    public boolean getWinchBreaker() { return shootWinchBreaker; }
    public boolean getKickLimit() { return kickLimitValue; }
    public double getSuckerCurrent() { return (suckerMotor.getOutputCurrent() + suckerMotor2.getOutputCurrent()) / 2.0; }


    static double clamp(double x)
    {
	if (x > 1.0) return 1.0;
	if (x < -1.0) return -1.0;
	return x;
    }

    public void stopMotors()
    {
	setSuckerPower(0);
	setWinchShootPower(0);
	//setWinchHangPower(0);
	setFLPower(0);
	setBLPower(0);
	setFRPower(0);
	setBRPower(0);
    }

    public void initAutonomous(PyroMain main)
    {

    }
    public void initOperator(PyroMain main)
    {

    }
    public void runAutonomous(PyroMain main)
    {
        updateValues(main);
    }
    public void runOperator(PyroMain main)
    {
        updateValues(main);
    }
    
    public void runDiagnostics(PyroMain main)
    {
        updateValues(main);
    }

    private void updateValues(PyroMain main)
    {
        dsLCD.println(DriverStationLCD.Line.kUser2, 1, "DIAG=".concat(String.valueOf(main.isDiagnosticsModeEnabled())));
        //dsLCD.println(DriverStationLCD.Line.kUser3, 1, "AUTO=".concat(String.valueOf(main.control.getAutoMode())));
        dsLCD.println(DriverStationLCD.Line.kUser3, 1, "TRIGGER=".concat(String.valueOf(hStick.getTrigger())));
        dsLCD.updateLCD();

        kickLimitValue = kickLimit.get();
        frMotor.set(-clamp(frMotorPower));
        flMotor.set(clamp(flMotorPower));
        brMotor.set(-clamp(brMotorPower));
        blMotor.set(clamp(blMotorPower));

        /*tmr = 0;
        if (tmr++ > 10)
        {
                tmr = 0;
                
        }*/

        suckerMotor.set(suckerPower);
        suckerMotor2.set(suckerPower);

        //if (getShooterCurrent() > shootWinchBreakerCurrent) { shootWinchBreaker = true; System.out.println("breaker tripped\n"); }
        if (winchShootPower > 0)
                shootWinchBreaker = false;

        //shootWinchBreaker = false; // debug

        if (!shootWinchBreaker) {
                winchShootMotor.set(winchShootPower);
        } else { // "breaker" tripped
                if (winchShootPower > 0) // positive is OK when tripped
                        winchShootMotor.set(winchShootPower);
                else
                        winchShootMotor.set(0.0);
        }        
        kickerSolenoid.set(kickerValue);
        kickerForwardSolenoid.set(kickerForwardValue);
    }

    public void setFRPower(double pwr) { frMotorPower = pwr; }
    public void setFLPower(double pwr) { flMotorPower = pwr; }
    public void setBRPower(double pwr) { brMotorPower = pwr; }
    public void setBLPower(double pwr) { blMotorPower = pwr; }

    public void setSuckerPower(double pwr)
    {
        suckerPower = -pwr;
    }

    public void setWinchShootPower(double pwr) { winchShootPower = pwr; }
    public void setKickerLatch(boolean value) { kickerValue = value; }
    public boolean getKickerLatch() { return kickerValue; }

    public void setKickerForwardValue(boolean value) { kickerForwardValue = value; }


    public double getFRPower() { return -frMotor.get(); }
    public double getFLPower() { return flMotor.get(); }
    public double getBRPower() { return -brMotor.get(); }
    public double getBLPower() { return blMotor.get(); }
    public double getSuckerPower() {return suckerMotor.get(); }
    public double getWinchShootPower() { return /*winchShootMotor.get();*/ 0; }

    public double getNewFRPower() { return frMotorPower; }
    public double getNewFLPower() { return flMotorPower; }
    public double getNewBRPower() { return brMotorPower; }
    public double getNewBLPower() { return blMotorPower; }
    public double getNewSuckerPower() {return suckerPower; }
    public double getNewWinchShootPower() {return winchShootPower; }
    public boolean getNewKickerValue() {return kickerValue; }
    //Relay::Value getNewWinchBrakeValue() {return winchBrakeValue;}

    // joystick Y axis thought to require negation
    public double getForward() {return -lStick.getY();}
    public double getStrafe() {return lStick.getX();}
    public double getTurn() {return rStick.getX();}
    public boolean getLeftTrigger() {return lStick.getTrigger();}
    public boolean getRightTrigger() {return rStick.getTrigger();}

    /*public int getFREncoder() { return frEncoder.get(); }
    public int getFLEncoder() { return flEncoder.get(); }
    public int getBREncoder() { return brEncoder.get(); }
    public int getBLEncoder() { return blEncoder.get(); }*/

    //temporary
    public boolean getOverBumpButton() {return false; }
    public boolean getWinchHangButton() {return hStick.getRawButton(4); }
    public boolean getKickerButton() {return hStick.getRawButton(2); }
    public boolean getScriptAbortButton() {return hStick.getRawButton(3); }

    public double getShooterCurrent() {
        return winchShootMotor.getOutputCurrent();
    }

    //added by rohan for diagnostic purposes
    /*public double getWinchFPCurrent() { return winchHangMotor.getOutputCurrent(); return 0; }*/
    /*public double getWinchCIMCurrent() { return winchHangMotor2.getOutputCurrent(); return 0; }*/
    //public double getArmServoAngle() {return armServo.getAngle();}

    //public double getGyroValue() { return 0; } // { return gyroValue; }
    //public void resetGyro() { } // { gyro.reset(); }
    /*public void resetEncoders()
    {
        frEncoder.reset();
        flEncoder.reset();
        brEncoder.reset();
        blEncoder.reset();
    }*/

}

File: svn/tags/game/Spy-Pyro/HAL.cpp
#include "HAL.h"
#include "WPILib.h"
#include "locator.h"
#include "control.h"

HAL* const hal = new HAL();

static float clamp(float x)
{
	if (x > 1.0) return 1.0;
	if (x < -1.0) return -1.0;
	return x;
}

void HAL::initAutonomous(void)
{
	//compressor.Start();
	#if 0
	frEncoder.Reset();
	flEncoder.Reset();
	brEncoder.Reset();
	blEncoder.Reset();
	gyro.Reset();
#endif
}

void HAL::stopMotors()
{
	setSuckerPower(0.f);
	setWinchShootPower(0.f);
	setWinchHangPower(0.f);
	setFLPower(0.f);
	setBLPower(0.f);
	setFRPower(0.f);
	setBRPower(0.f);
}

void HAL::initOperator(void)
{
	//compressor.Start();
#if 0
	frEncoder.Reset();
	flEncoder.Reset();
	brEncoder.Reset();
	blEncoder.Reset();
	gyro.Reset();
#endif
}

void HAL::runAutonomous(void)
{
	updateValues();
}

void HAL::runOperator(void)
{
	updateValues();
}

void HAL::runDiagnostics(void)
{
	updateValues();
}


void HAL::updateValues()
{
	dsLCD->Printf(DriverStationLCD::kUser_Line1, 1, "DIAG=%d", (int) isDiagnosticsModeEnabled());
	dsLCD->Printf(DriverStationLCD::kUser_Line2, 1, "AUTO=%d", (int) control->getAutoMode());
	dsLCD->UpdateLCD();

	kickLimitValue = kickLimit.Get();
	//kickLimitValue = kickerAnalogChannel.GetAverageValue() > 250;
	//Right motors currently thought to require negation. Subject to change if found incorrect.
	//Wait(0.1);
	//printf("%f\t%f\t%f\t%f\n",frMotorPower,flMotorPower,brMotorPower,blMotorPower);
#if 1
	frMotor.Set(-clamp(frMotorPower));
	flMotor.Set(clamp(flMotorPower));
	brMotor.Set(-clamp(brMotorPower));
	blMotor.Set(clamp(blMotorPower));
#endif
	
	//return;
	static int tmr = 0;
	if (tmr++ > 10)
	{
		tmr = 0;
		//printf("kickLimit = %i\n",(int)kickLimitValue);
		//printf("suckerPower = %f\n",suckerPower);
		//printf("GetVoltage() = %f\nGetValue() = %d\nGetAverageValue() = %d\nGetAverageVoltage() = %f\n\n",
		//		kickerAnalogChannel.GetVoltage(), (int)kickerAnalogChannel.GetValue(),
		//		kickerAnalogChannel.GetAverageValue(), kickerAnalogChannel.GetAverageVoltage());
		//printf("hangerWinch = %f brake = %i\n",winchHangPower,(int)hangerBreakValue);
		//printf("updating values: %f %f %f %f\n",frMotorPower,flMotorPower,brMotorPower,blMotorPower);
		//printf("winch = %f\n",winchHangPower);
		//printf("solenoid = %i",armValue);
		//printf("enc: fr = %i fl = %i br = %i bl = %i\n",frEncoder.Get(),flEncoder.Get(),brEncoder.Get(),blEncoder.Get());
		//float x = winchShootMotor.GetOutputCurrent();
		//if (x != 0)
		//printf("%f %f %f\n",locator->getX(), locator->getY(), locator->getTheta());
		//	printf("%f\n",x);
		//printf("solenoid latch = %i\n",kickerValue);
	}
	//return;

#if 1
	suckerMotor.Set(suckerPower);
	suckerMotor2.Set(suckerPower);
#endif
	//armServo.SetAngle(armServoAngle);
	
#if 1
	// breaker control
	if (getShooterCurrent() > shootWinchBreakerCurrent) { shootWinchBreaker = true; printf("breaker tripped\n"); }
	if (winchShootPower > 0)
		shootWinchBreaker = false;
	
	//shootWinchBreaker = false; // debug
	
	if (!shootWinchBreaker) {
		winchShootMotor.Set(winchShootPower);
	} else { // "breaker" tripped
		if (winchShootPower > 0) // positive is OK when tripped
			winchShootMotor.Set(winchShootPower);
		else
			winchShootMotor.Set(0.0);
	}
	
	
	//winchHangMotor.Set(winchHangPower);
#ifdef SPY
	//winchHangMotor2.Set(-winchHangPower);
#else
	//const float kCurrentLimit = 1;
	//winchHangMotor2.Set(winchHangPower * kCurrentLimit);
#endif
#endif
	// Arm
	//armSolenoid.Set(armValue);
	kickerSolenoid.Set(kickerValue);

	kickerForwardSolenoid.Set(kickerForwardValue);
	//hangerBreak.Set(!hangerBreakValue);

	//winchBrake.Set(winchBrakeValue);
#if 0
	gyroValue = gyro.GetAngle();
#endif
}

//Positive should be clockwise. Inversion happens at update.
void HAL::setFRPower(float pwr) { frMotorPower = pwr; }
void HAL::setFLPower(float pwr) { flMotorPower = pwr; }
void HAL::setBRPower(float pwr) { brMotorPower = pwr; }
void HAL::setBLPower(float pwr) { blMotorPower = pwr; }
void HAL::setSuckerPower(float pwr)
{
#ifdef PYRO
	suckerPower = -pwr;
#else
	suckerPower = pwr;
#endif
}
void HAL::setWinchShootPower(float pwr) { winchShootPower = pwr; }
void HAL::setKickerLatch(bool value) { kickerValue = value; }
bool HAL::getKickerLatch() { return kickerValue; }
void HAL::setWinchHangPower(float pwr)
{
	//printf("setting winch power to %f",pwr);
	if (!hangerBreakValue) {
		winchHangPower = pwr;
	//	printf(", success\n");
	} else {
	//	printf(", fail\n");
	}
}

void HAL::setKickerForwardValue(bool value) { kickerForwardValue = value; }
void HAL::setArm(bool value) { armValue = value;}
void HAL::setArmServoAngle(float angle){ armServoAngle = angle;}
void HAL::setHangerBreak(bool value)
{
#ifdef PYRO
	hangerBreakValue = value;
	//setWinchHangPower(0.f); // who the FUCK wrote this?
#endif
}

bool HAL::getArm() { return armValue; }
//void HAL::setWinchBrake(Relay::Value value) {winchBrakeValue = value;}

//Change these if right motors are found to be positive=forward.
float HAL::getFRPower() { return -frMotor.Get(); }
float HAL::getFLPower() { return flMotor.Get(); }
float HAL::getBRPower() { return -brMotor.Get(); }
float HAL::getBLPower() { return blMotor.Get(); }
float HAL::getSuckerPower() {return suckerMotor.Get(); }
float HAL::getWinchShootPower() { return /*winchShootMotor.Get();*/ 0; }
float HAL::getWinchHangPower() { /*return winchHangMotor.Get();*/ return 0; }

float HAL::getNewFRPower() { return frMotorPower; }
float HAL::getNewFLPower() { return flMotorPower; }
float HAL::getNewBRPower() { return brMotorPower; }
float HAL::getNewBLPower() { return blMotorPower; }
float HAL::getNewSuckerPower() {return suckerPower; }
float HAL::getNewWinchShootPower() {return winchShootPower; }
bool HAL::getNewKickerValue() {return kickerValue; }
float HAL::getNewWinchHangPower() {return winchHangPower;}
//Relay::Value HAL::getNewWinchBrakeValue() {return winchBrakeValue;}

// joystick Y axis thought to require negation
float HAL::getForward() {return -lStick.GetY();}
float HAL::getStrafe() {return lStick.GetX();}
float HAL::getTurn() {return rStick.GetX();}
bool HAL::getLeftTrigger() {return lStick.GetTrigger();}
bool HAL::getRightTrigger() {return rStick.GetTrigger();}

int HAL::getFREncoder() { return frEncoder.Get(); }
int HAL::getFLEncoder() { return flEncoder.Get(); }
int HAL::getBREncoder() { return brEncoder.Get(); }
int HAL::getBLEncoder() { return blEncoder.Get(); }

#if 0
float HAL::getGyroValue() { return gyroValue; }
void HAL::resetGyro() { gyro.Reset(); }
void HAL::resetGyro() {
	gyro.Reset();
	}
#endif

//temporary
bool HAL::getOverBumpButton() {return false; }
bool HAL::getWinchHangButton() {return hStick.GetRawButton(4); }
bool HAL::getKickerButton() {return hStick.GetRawButton(2); }
bool HAL::getScriptAbortButton() {return hStick.GetRawButton(3); }

float HAL::getShooterCurrent() {
#if 1
	return winchShootMotor.GetOutputCurrent();
#else
	return 0;
#endif
}

//added by rohan for diagnostic purposes
float HAL::getWinchFPCurrent() { /*return winchHangMotor.GetOutputCurrent();*/ return 0; }
float HAL::getWinchCIMCurrent() { /*return winchHangMotor2.GetOutputCurrent();*/ return 0; }
float HAL::getArmServoAngle() {return armServo.GetAngle();}

float HAL::getGyroValue() { return 0; } // { return gyroValue; }
void HAL::resetGyro() { } // { gyro.Reset(); }
void HAL::resetEncoders()
{
	frEncoder.Reset();
	flEncoder.Reset();
	brEncoder.Reset();
	blEncoder.Reset();
}


